Add new  API server and enhance codes

longxice opened the pull request:
Secret info leak on frontend, it should be in server All apis should be through the server
- Implement API server,

Fix bugs
- Fixed responsive logo

Enhancements
- Align center for app gif in the readme file

Commit:
Add new  API server and enhance codes

Secret info leak on frontend, it should be in server
All apis should be through the server
- Implement API server,

Fix bugs
- Fixed responsive logo

Enhancements
- Align center for app gif in the readme file

anhldbk commented on server.js:
```diff
@@ -1,8 +1,12 @@
 const morgan = require("morgan");
 const { uuid } = require("uuidv4");
+const crypto = require("crypto-js");
+const NodeRSA = require("node-rsa");
```
@longxice naming style: why not `nodeRSA` ? Or shorter, `rsa`?

longxice replied:
@anhldbk I think it should be 'rsa', because this is a Node.js application, so it should not include 'node' or any JavaScript-related words.

anhldbk replied:
@longxice ok. Go ahead.

leonthaibao commented on server.js:
Should we put this file into a folder? maybe a `server` folder?

longxice replied:
@leonthaibao 
If the file contains server-side code or configuration files for a web application or server, it's generally a good practice to keep them in a specific folder, such as a server folder, to keep the project organized and maintainable. Keeping them in a separate folder also helps to ensure that they are not accidentally exposed to the public or served as static assets, which could potentially compromise the security of the application.

However, if the file is a client-side script or a standalone utility script, it may not be necessary to put it in a specific folder as long as it's located in a place where it can be easily found and accessed by the developer or user who needs it.
The server is standalone and simple. Let me keep it at the project root directory.

leonthaibao commented on server.js:
```diff
@@ -14,118 +18,227 @@ app.use(express.json());
+      },
+      body: JSON.stringify({ app_id: appId, payment_id: paymentId, time, mac })
+    });
+    return res.json(await response.json());
```
Why do we use `await` before `response.json()`?

longxice replied:
@leonthaibao 
The response.json() method returns a Promise that resolves to a JavaScript object representing the parsed JSON data from the response body. By using the await keyword before calling this method, we tell the JavaScript engine to pause execution of the current function until the Promise is resolved, and then assign the resulting value to a variable.

Without using await, the code would continue executing immediately after calling response.json(), before the Promise has resolved, which could result in the function attempting to use the parsed data before it is available. This could lead to errors or unexpected behavior in the program.

leonthaibao commented on src/app/util/helpers.js:
```diff
@@ -4,5 +4,26 @@ module.exports = {
+        }
+        return "";
+    },
+    validateAmount: function (amount) {
```
We should compare the balance and the amount too. If the balance only has 100 but the amount is 101, we should throw an error
The amount field should have a min value like 0 or 1000

longxice replied:
@leonthaibao It will add balance check case soon

longxice replied:
@leonthaibao The minimum amount that can be transferred is 1 VND

leonthaibao commented on src/features/request/PayrollRequest.js:
```diff
@@ -146,11 +121,11 @@ export const PayrollRequestContainer = () => {
                     type="text"
                     className="payment-form-field-input, amount"
-                    placeholder="1000xxx"
+                    placeholder="Enter amount"
```
![Screenshot 2023-04-24 at 11 11 39](https://user-images.githubusercontent.com/45265127/233898718-dd80c6e0-209c-415d-b11b-f7b3cc43b561.png)
I found there is an error when the amount is over 1 million, pls check this

longxice replied:
@leonthaibao 
OK let me check. Thanks!

anhldbk commented on server.js:
```diff
@@ -14,118 +18,227 @@ app.use(express.json());
 });
 
+const isMock = process.env.REACT_APP_IS_MOCK === "1";
+const SANDBOX_URL = process.env.REACT_APP_SANDBOX_URL;
```
@longxice inconsistent style btw constants, e.g: SANDBOX_URL vs isMock
@leonthaibao what is your suggestion?

longxice replied:
@anhldbk For consistency, I will change it, from isMock to IS_MOCK

anhldbk commented on src/app/util/helpers.js:
```diff
@@ -4,5 +4,26 @@ module.exports = {
         parts[0] = parts[0].replace(/\B(?=(\d{3})+(?!\d))/g, ',');
         return parts.join('.');
+    },
+    validatePhone: function (phoneNo) {
```
@longxice pls add tests with Jest: https://jestjs.io/docs/getting-started

longxice replied:
@anhldbk Yes I will do

leonthaibao commented on README.md:
In the usage section, it only has the instruction to run the server side, we should add the instruction to run the client side too

longxice replied:
@leonthaibao It is a problem if users who are not familiar with Node.js cannot start the client. Yes, I will address this issue.

Commit:
Fix MR reviews

- NodeRSA -> rsa
- isMock -> IS_MOCK
- Add test case for helpers
- Wrong format when value of amount is over 1 million
- Add guideline how to start the application (frontend) in the read me file.

longxice commented:
@anhldbk , @leonthaibao all things are fixed. Thank your for your comments.

anhldbk commented on server.js:
```diff
@@ -14,118 +18,227 @@ app.use(express.json());
+        sub_return_code: 1,
+        sub_return_message: "Success",
+        data: {
+          name: "Longhb",
```
@longxice pls use a more popular name.

longxice replied:
@anhldbk This is fixed

Commit:
Fix MR reviews

- Update the mock data to include the full name

leonthaibao reviewed (approved):
lgtm
